ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,124 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+class ImpendingTimeout(Exception):
+    pass
+
+
+def crash_flush(_, __):
+    """Raises an `ImpendingTimeout` error which causes the tracer to shutdown."""
+    # TODO: Update to tag error in local root parent instead of raising the error.
+
+    raise ImpendingTimeout("Datadog detected an Impending Timeout")
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        remaining_time_in_millis = apm_flush_deadline
+
+    remaining_time_in_seconds = (remaining_time_in_millis - 100) / 1000
+    signal.setitimer(signal.ITIMER_REAL, remaining_time_in_seconds, 0)
+
+
+def datadog_instrumentation(func, args, kwargs):
+    """Patches an AWS Lambda handler function for Datadog instrumentation."""
+    context = args[-1]  # context is always the last parameter
+    check_timeout(context)
+
+    return func(*args, **kwargs)
+
+
+def modify_module_name(module_name):
+    """Returns a valid modified module to get imported."""
+    return ".".join(module_name.split("/"))
+
+
+def get_handler_and_module():
+    """Returns the user AWS Lambda handler and module."""
+    path = os.environ.get("DD_LAMBDA_HANDLER", None)
+    if path is None:
+        from datadog_lambda.wrapper import datadog_lambda_wrapper
+
+        handler = getattr(datadog_lambda_wrapper, "__call__")
+
+        def wrapper(func, args, kwargs):
+            return datadog_instrumentation(func, args, kwargs)
+
+        return handler, datadog_lambda_wrapper, wrapper
+    else:
+        parts = path.rsplit(".", 1)
+        (mod_name, handler_name) = parts
+        modified_mod_name = modify_module_name(mod_name)
+        handler_module = import_module(modified_mod_name)
```

Comment:
If a user sets DD_LAMBDA_HANDLER to an invalid path this import will raise an error. Should we catch and handle the exception here?

Reply:
This is already checked in `datadog_lambda` package, which will indeed be updated once we stop patching and move the tracing features here. I didn't add it because of yagni.